Sdx.Web.Url 用ユニットテストの修正

sd-tomita opened the pull request:
## 概要
#21 のつづきです。

UnitTest にまだテストできていなかったケースをいくつか追加して、
必要に応じて実装側を修正するためのPRです。
## TODO
- [x] `TestMethod1()` の名前を意味のあるものに変更(テンプレートのままなので)
- [x] プロパティの記述方法変更([**フィールドの自動実装**](https://msdn.microsoft.com/ja-jp/library/bb384054.aspx) を使う)
- [x] テストケースの追加
  - 全テストに追加する
    - パラメータを追加(set)する
    - Buildの引数にパラメータ(Dictionary)を渡す
    - Buildの引数にパラメータ(Array)を渡す
    - 空のDictionary、key しかないDictionary を渡す
    - まだパラメータが何もセットされていない状態の Sdx.Web.Url に array や List を渡す
- [x] 他にも必要そうなテストが無いか検討する
- [x] 前項で決定したテストの追加
  - 詳細： https://github.com/SunriseDigital/cs-sdx/pull/24#issuecomment-142814610
    - 存在しない値を取得
    - 配列形式のクエリ文字列をセット
    - 値の無いクエリ文字列をセット(前述では `?key=` だったので、`?key` を試してみる)
- [x] 追加したテストに伴うコードの修正(必要に応じて適宜)
- [x] Sdx.Web.Url.Param の private 化
  - Get, Set メソッドを使うようにする。
- [x] ~~Sdx.Web.Url.Param を `Dictionary<string,string>` → `Dictionary<string,object>` に変更する。(同じキーで複数の値を持てるようにする)~~ Listで管理するように変更するためなし
- [x] Dictionary は追加した順番で取得できることは保証されていないため、List を使って保存するように変更する。
- [x] Get メソッドを配列、文字列の2パターン用意
- [x] Build メソッド周りの修正
- [x] 各 Get メソッドの挙動を検討する
- [x]  パラメータのリストを `List<KeyValuePair<string,string>>` で管理するように変更する
- [x] Set メソッドを上書き動作に修正
- [x] 追加のみを行うAddメソッドを別途作成する
- [x] パラメータを削除するメソッドも作成する
- [x] GetParam() は最初の要素の値を返すようにする
- [x] 特定のKeyに一致するValueの数を控えておき、Listで返すメソッド(GetParams)の中で必要な数だけループを行うようにする。
- [x] `List<KeyValuePair<string,string>>` を `List<Tuple<string,string>>` に変更

sd-tomita committed:
テスト用メソッド名が雛形のままになっていた箇所を修正

すみません漏れてました。

sd-tomita committed:
プロパティを自動実装に書き換えました

http://ufcpp.net/study/csharp/oo_property.html#auto
すごい、メンバ変数が全部いらなくなった。。。

sd-tomita committed:
パラメータの追加、削除を行うテストを追加

sd-tomita committed:
空のDictionary、値が空のDictionaryをセットしてみるテストを追加

念のため 空文字 と null の2パターン用意

sd-tomita committed:
存在しないキーを削除しようとするテストを追加

何もしない という動作を想定。
指定されたキーは存在しないことをエラーなどで返すことも考えましたが、
パラメータの内容はライブラリの使用者がセットしたものだけが全てなので
今回はわざわざ存在しないことをエラーで通知する必要は無いかと考えました。

sd-tomita commented:
## 思案中

他にやっておいたほうがよさそうなテストや、あったほうが良い機能など考えています。
とりあえず今思いついているやつ。~~(他にあったら随時追記予定)~~ (とりあえずここにはもう足さない。何かあれば別途コメントに書く。)
- 存在しない値を取得してみる
  - `Assert.Equal(null, url.Param["unknown"]);` とかそんなかんじ。
- ~~クエリ文字列だけを取得。やるとしたら `BuildQueryString` を public にすれば事は足りそう。。。と書いてはみたものの、そもそもこの機能必要あるかな？とりあえずメモだけ残して、作業するかどうかはまだ保留。~~ しないことになりました。
- 配列系のクエリ文字列を使ったテスト。なんとなくコケそう。
  - 普通の：`?array[]=value0&array[]=value1&array[]=value2`
  - 連想配列：`?list[key0]=value0&list[key1]=value1&list[key2]=value2`
- 値の無いパラメータ(=すら付いていない。)
  - 例：`?novalue&novalue2&key=value&novalue3`
- `Build()` の引数に渡すDictionary、List、Arrayの要素を複数に増やしてみる。大丈夫そうだけど一応やっておこう。

sd-tomita committed:
存在しない値を取得するテストを追加

現状は例外になる。

gomo commented:
> やるとしたら BuildQueryString を public にすれば事は足りそう

テストのために隠せるメンバーを公開するのって微妙ですよね。ただ、UnitTestが別のアッセンブリーに属すので、どうしてもPublicにしないとダメな時有りますけどね。かと言って、Sdxにテスト入れちゃうのもなあ。ひつようのないものをメモリに読み込むわけだし・・・

今回はあまり必要無さそうな機能なのでとりあえず足さなくて良いかな。

sd-tomita commented:
@gomo 
了解しました、今回はひとまず見送ります。

> テストのために隠せるメンバーを公開するのって微妙ですよね。

そうなんですよね。思いついてはみたもののクエリ文字列だけが欲しいって状況ってほぼなさそうですし、 public にするにはちょっと割りに合わない気がします。

sd-tomita commented:
## メモ

このテストを行うと、存在しないキーにアクセスしている時点で、
System.Collections.Generic.KeyNotFoundException が発生する。

``` C#
Assert.Equal(null, url.Param["unknown"]);
```

Sdx.Web.Url.Param プロパティが返しているのは、あくまでも Dictionary オブジェクト自体なので、Param["unknown"] でアクセスされた場合に例外を返さないようにするには、自分でインデクサ周りをどうにかするしかない。

ちなみに Dictionary のインデクサの挙動はこのへん。だいたいはItemプロパティに実装されているらしい。
https://msdn.microsoft.com/ja-jp/library/zyxt2e2h(v=vs.100).aspx

Dictionaryは継承不可なので、やるなら自分でIDictionary インターフェースを実装するクラスを作るとかそんなかんじになるのか。

ただ、存在しないキーでアクセスした場合に例外になること自体は真っ当な動作だと思うのでインデクサに手を入れて例外にならないようにするのはどうなのかという気もする。となると、単純に Get メソッド作るほうが手っ取り早いかもしれない。

ちなみにTryGetValueメソッドなどを使って、クライアントコード側でチェックするのが、Dictionaryから値を取り出す際のセオリーのよう。これだとこのテスト自体が間違っていることになるのか＼(^o^)／
https://msdn.microsoft.com/ja-jp/library/9tee9ht2.aspx

``` C#
string value = null;//TryGetValue で値が見つからない場合、この値がそのままデフォルト値になる
url.Param.TryGetValue("unknown", out value);
Assert.Equal(null, value);
```

sd-tomita commented:
## メモ2

https://github.com/SunriseDigital/cs-sdx/pull/24#issuecomment-142889708
↑の話。よく考えたら、このまま例外になってていいかも。
存在しないキーを指定したときに null が返ってきてしまうと、value がもともと null のパラメータと区別がつかなくなってしまう。少なくとも null を返すべきではない。

となると、このまま例外にしておいたほうがはっきりしてていいのかもしれない、とふと思った。
ジェネリックのDictionary が存在しないキーを指定された際に null ではなく例外をかえすようになったのもそういう背景もあるのかもしれない(わからないけど)。

sd-tomita committed:
存在しないキーを指定した場合のテストは一旦中止

存在しないキーを指定した場合、例外になるほうが正しい挙動に思えたため一旦コメントアウト。
ここで null が返ってくると、もともと値が null のパラメータと区別がつかなくなる。

gomo commented:
> ↑の話。よく考えたら、このまま例外になってていいかも。

ですかね。私はよく

``` c#
if(dic.ContainsKey("unknown"))
{
```

でチェックしてます。

どうしても例外にしたくなかったら、GetParamメソッドを作るかParam自体を独自クラスにするかですね。

> 自分でIDictionary インターフェースを実装するクラスを作る

はどういう利点があるのかな？まあ、辞書っぽい何かならつけておくと、便利になることはあるかも。例えば、Sdx.Diagnostics.Debug.Dump()でうまくDUMPしてもらえるとか。

> ジェネリックのDictionary が存在しないキーを指定された際に null ではなく例外をかえすようになったのもそういう背景もあるのかもしれない(わからないけど)。

これは、単純に、NULLがきても結局その先でヌルポ（C#だと違う例外だった気がするけど）が出るので、早いうちに停めておいたほうがいいからだと思います。

> 存在しないキーを指定した場合のテストは一旦中止 …

ちなみに、例外が出るかどうかテストをするというのもあります。今回は「Dictinaryの存在しないキーにアクセスしたら例外が出る」というのはプログラマーが知ってることなので特に必要ないですが、「これをすると例外になるのね」という確認になるのでわざわざそういうテストを追加することは有ります。

https://github.com/SunriseDigital/cs-sdx/blob/proj/ormUpdate/UnitTest/DbRecordTest.cs#L516-L522

sd-tomita commented:
@gomo 

> どうしても例外にしたくなかったら、GetParamメソッドを作るかParam自体を独自クラスにするかですね。

ですかね。GetParamメソッドを作っても、結局 `Param["key"]` でアクセスされたら同じなので、どうしても例外にしたくない場合は後者の案がよさそうな気がします。今回はコメントのとおり例外のままにしようと思いますが、今後何かあれば参考にいたします:innocent:ありがとうございます。

> ちなみに、例外が出るかどうかテストをするというのもあります。今回は「Dictinaryの存在しないキーにアクセスしたら例外が出る」というのはプログラマーが知ってることなので特に必要ないですが、「これをすると例外になるのね」という確認になるのでわざわざそういうテストを追加することは有ります。

なるほど、そういうテストをすることもあるんですね。あとでさっそく追加してみます！
↓このコメントにも通づるところがありますね。
https://github.com/SunriseDigital/cs-sdx/pull/21#discussion_r39821483

sd-tomita committed:
期待した例外が出るかどうかのテストを追加

sd-tomita committed:
配列、連想配列のクエリ文字列を使ったテストメソッドを追加

テスト動作自体はこれから足します

sd-tomita commented:
## 配列含みのクエリ文字列について

どういうインターフェースにしようか考え中です。
個人的には使う側からしたらこれまでと同じように `url.Param` から取得できるほうが望ましい気はします。
ただ、key の型がこれまでと異なるので、そのへんをどうにかする作業は必要そうです。
(下記コードは検討中のテスト内容です。 Assert.Equal で引数同士の型が合わずコンパイルエラーになるので、とりあえずコミットせずにメモだけ残したものです)

``` C#
[Fact]
public void TestArrayParams()
{
  var url = new Sdx.Web.Url("http://example.com/path/to/api?array[]=value0&array[]=value1&array[]=value2");

  var arr = new string[]{ "value0", "value1", "value2" };
  Assert.Equal(arr, url.Param["array"]);
  Assert.Equal("value0", url.Param["array"][0]);
}

[Fact]
public void TestHashParams()
{
  var url = new Sdx.Web.Url("http://example.com/path/to/api?dic[key0]=value0&dic[key1]=value1&dic[key2]=value2");

  var dic = new Dictionary<string,string>(){ { "key0", "value0" },{ "key1", "value1" },{ "key2", "value2" } };
  Assert.Equal(dic, url.Param["list"]);
  Assert.Equal("value0", url.Param["dic"][0]);
}
```

sd-tomita commented:
## メモ

※gomo さんと相談した内容コミで忘れないうちに書き留めています。
- `url.Param<string, string>` はやめる。
  - このプロパティは private にする。
  - 同じキーの値を複数持てるように、`url.Param<string, object>` に変更する。
- 値の取得はメソッドで行うようにする。
  - 配列と文字列それぞれの取得メソッドを用意して、使う側で選べるようにする。
  - Param プロパティが private になるので追加用メソッドも用意する。
- 配列はそのまま返す。文字列指定の場合、配列から最後の要素を返す。
  - (存在するキーに対して新しい値をセットした場合、上書きされるのがこれまでの想定挙動なので、最後に追加された値が返るようにすればライブラリの使用側から見れば上書きと同じになる)
- 配列の要素が必ず追加順に並んでいることが保証されているかどうかは要確認。(たぶん保証されていない気がする)

gomo commented:
> 配列と文字列それぞれの取得メソッドを用意して、使う側で選べるようにする。

それぞれ、取得した後そのオブジェクトを変更したらどうなるのか？どうなるのが相応しいのか考えながらテスト、実装してくださいね。

gomo commented:
> あとでさっそく追加してみます！

は必要ないです。

> 今回は「Dictinaryの存在しないキーにアクセスしたら例外が出る」というのはプログラマーが知ってることなので特に必要ないですが

ですので。

ただ、[この辺](https://github.com/SunriseDigital/cs-sdx/pull/24#issuecomment-143179523)の変更であったほうが良いのか、無くてもOKなのか・・・考えてみてください。

sd-tomita commented:
@gomo 
- > それぞれ、取得した後そのオブジェクトを変更したらどうなるのか？どうなるのが相応しいのか考えながらテスト、実装してくださいね。
- > ただ、[この辺](https://github.com/SunriseDigital/cs-sdx/pull/24#issuecomment-143179523)の変更であったほうが良いのか、無くてもOKなのか・・・考えてみてください。

諸々了解ですー
作業しつつ考えてみます。

gomo commented:
## 想定すべきややこしいパターン

### 値の上書き

``` c#
var url = new Sdx.Web.Url("http://example.com/path/to/api?array[]=value0&array[]=value1&array[]=value2");

var arr = url.GetParams('array[]');
arr.Add("value3");

var urlString = url.Build();
```

``` c#
var url = new Sdx.Web.Url("http://example.com/path/to/api?array[]=value0&array[]=value1&array[]=value2");

var str = url.GetParam('array[]');
str = "foobar";

var urlString = url.Build();
```

#### そもそも参照型の代入って？

``` c#
var a = "bar";
var b = a;
b = "foo";

//aの中身は？
```

#### そもそも参照型のオブジェクトの変更

``` c#
var a = new List<string>{"aaa", "bbb"};
var b = a;
b = b.Add("ccc");

//aの中身は？
```

### こんなURLの時値は？

http://example.com/path/to/api?foo=&bar=hoge

``` c#
var str = url.GetParam("foo");
```

http://example.com/path/to/api?foo=&foo=

``` c#
var arr = url.GetParams("foo");
```

gomo commented:
```c#
var a = "bar";
var b = a;
var b = "foo";

//aの中身は？
```

sd-tomita committed:
Sdx.Web.Url.Param を private なプロパティにするための修正

とりあえず従来の Dictionary<string,string> のまま、Getメソッド、Setメソッドに
各種を差し替えるところまで。

sd-tomita committed:
配列パラメータ用のテストケース追加

sd-tomita committed:
連想配列用のテストを追加

C#ではクエリ文字列で [] をキーに添えても配列が表現できるわけではないので
連想配列でデータを送信する場面がそもそもないかも。このテストはいらなそうだったら削除予定。

sd-tomita committed:
不要なテストメソッドを削除

7925fcc でいらなそうだったら削除予定と書いていますが
やはり不要になりましたｗ

sd-tomita commented:
## Dictionary は並び順を保持していない

テスト上は特に順番が変わったりすることは今のところありませんが、今後もそれが保証されているわけではないようです。
http://ufcpp.net/study/dotnet/bcl_collection.html#Dictionary
http://www.atmarkit.co.jp/fdotnet/dotnettips/429dictionary/dictionary.html

とりあえず Dictionary はやめたほうがよさそうです。

## 対応案(検討中)

List ならインデックス順で保証できるので List を使うのが良さそうかと考えています。
どう使うのがよさそうかは検討中です。

sd-tomita commented:
### List でパラメータを管理する方法案のメモです(修正・追記あり)
1. key と value のそれぞれの List を作る
2. ~~key だけを格納した List を別に作る。(これまで使っていた Dictonary はそのまま)~~この案は、同一キー間での順番を維持できないのでナシ。(例えば `?key=val1&key=val2&key=val3` というパラメータをコンストラクターに渡しても、キーだけの並び順しか保持していないので、Build 時に `?key=val2&key=val1&key=val3` とかなる可能性がある。) 
3. `List<Dictionary<string,string>>`

どの方法でも同じキーが複数あったときどうするかが問題。（List、string どちらで返す場合でも結局全要素見て検索するのであまり変わらない)

なお、以前に話のあった `Dictionary<string, List<string>>` ですが、例えば元のクエリが `?key=val1&foo=bar&key=val2` のように違う key をはさんでいたりすると、その順番までは維持できなくなるので、各キー・値は一対一で考えるのが無難に思えてきました。
Getメソッドで単体を取るなら後から追加したものを、Listで取るならそのときに同じキーの値をまとめてList化して返せば話は簡単そうです。

となると、1番目 or 3番目のどちらかかなと。
key と value を同じ変数に格納できて管理がしやすそうなので、今回は3番目の方法でやってみます。

sd-tomita committed:
テスト追加したので一旦コミット

まだ増えたり減ったりするかとおもいますが、一旦ここで記録を残しています

sd-tomita committed:
メソッド名をテスト内容に即した名前に変更

ここでテストする内容の本質は Array がどうこうって話ではなく、
同じキーのパラメータを複数持っている場合の挙動についてということなので
変更しました。

sd-tomita committed:
同一キーで異なる値を持つクエリに対応するための作業中

・パラメータを List<Dictionary<string,string>> で管理
・コンストラクター、Getメソッド、Setメソッドをパラメータ管理方法の変更に追従

sd-tomita committed:
テストした内容をコミットしていなかったのでコミット

とりあえず Get、Set メソッドの通常挙動のみ確認しています。

sd-tomita committed:
Buildメソッドの挙動を新しいパラメータ管理方法に追従しました

sd-tomita committed:
存在しないキーで値をGetしようとした際のテストを追加

そもそもパラメータを List<Dictionary<string,string>> で管理していることは
コードを使う側からは知る由もないので、テストがあったほうが良い

sd-tomita committed:
存在しないキーで値をGetしようとしたら例外を投げるようにしました

このままnewしただけのlistを返しても、このlistで何かしようとすれば
どの道例外が投げられるので、この段階で例外を投げています

sd-tomita committed:
First()を不要な箇所で使っていたので修正

ここはキーで決め打ちできるので変更しました。

sd-tomita committed:
コンパイルできるようになったのでテストのコメントアウトを解除

Paramの管理方法変更中はコンパイルができなくなっていたので
しばらくコメントアウトしていました。

sd-tomita committed:
同じキーの値を複数持てるようにしたことで挙動が変わるのでテスト追加

Setメソッドと、Build(Dictionary) で挙動が異なるようになる。
(今までは上書きのみだったが、Setは上書きしないようする必要ができたので
そのあたりもテストするため)

sd-tomita committed:
Build(Dictionary) は上書き動作になるよう修正

要は Build の動作は既存と同じままになります。

sd-tomita committed:
空のDictionaryをBuildの引数に渡すとテストがコケていたのを修正

sd-tomita committed:
空文字、null 絡みのテストが少し足りなかったので別途追加

gomo commented:
> List でパラメータを管理する方法案のメモです

同じキーを持つという性格上３しか無さそうですね。

``` c#
List<KeyValuePair<string, string>>
```

のほうが意図がはっきりするけど、KeyValuePairはstructです。どういう問題があると思います？

問題がなければKeyValuePairを使うがベストだと思います。

gomo commented:
``` c#
var kv1 = new KeyValuePair<string, List<string>>();
kv1.Key = "key1";
kv1.Value = new List<string>();
kv1.Value.Add("val1-1");

var kv2 = kv1;
kv1.Key = "key2";
kv2.Value.Add("val1-2");

//kv1.Keyは？
//kv1.Value.Countは？
```

tomita-ychr commented:
@gomo 
> KeyValuePairはstructです。どういう問題があると思います？

構造体について調べつつ考えていますが、
↓このへんでも出ているパフォーマンス的な問題でしょうか？正直ピンときていないですが16バイトってかなり少なそうです。
[クラスまたは構造体の選択](https://msdn.microsoft.com/ja-jp/library/ms229017.aspx)

sd-tomita commented:
@gomo 

> `List<KeyValuePair<string, string>>`

問題なさそうなのでこれで直してみますー。
(自分が書いたコメントを編集しようとしたら誤って削除してしまったので、口頭で終わっている話ではありますが、コメントに残します)

※昨日貼ってたリンク。問題がありうるとしたらメモリ的な話かとおもってました。
https://msdn.microsoft.com/ja-jp/library/ms229017.aspx

sd-tomita committed:
Param の型を List<KeyValuePair<string,string>> に変更

ひとまずプロパティの型と、コンストラクターのみ修正

sd-tomita committed:
ParamListの型変更に伴い、Setメソッドを修正

また、これまでのSetParamが追加の動作のみ行っていたので
別途AddParamメソッドを追加

動作としては以下のとおり。(おいおいメソッドのドキュメントに追記する)

SetParam･･･指定されたキーが存在している場合、古いものは全て削除してから追加する
AddParam･･･追加のみ行う

sd-tomita committed:
GetParamList だと、ParamListプロパティを丸ごと返すように見えるので名前変更

もし、ParamListを丸ごと返すような動作を追加するときの名前にも困るので
今のうちに名前を変えました

sd-tomita committed:
Get系メソッドの動作を現在のParamListの型に合わせて修正

sd-tomita commented:
## 作業リスト

もう始めてしまっているものもありますが、一応メモを残します。
- [x]  パラメータのリストを `List<KeyValuePair<string,string>>` で管理するように変更する
- [x] Set メソッドを上書き動作に修正
- [x] 追加のみを行うAddメソッドを別途作成する
- [x] パラメータを削除するメソッドも作成する
- [x] GetParam() は最初の要素の値を返すようにする
- [x] 特定のKeyに一致するValueの数を控えておき、Listで返すメソッド(GetParams)の中で必要な数だけループを行うようにする。

sd-tomita committed:
GetParam の動作がやや特殊なのでメソッドコメントで補足

最初の要素の値だけが返って来る という動作が
コードを使っているところだけ見てもパッとはわからないので。

sd-tomita committed:
パラメータの型やメソッドの動作変更に伴ってテストを修正

各テストに期待する結果も変わる箇所があるのでまずそこを修正。
新しいメソッドに対応したテストも追加

sd-tomita committed:
Build 系メソッドをParamListの型変更に動作を合わせました

sd-tomita committed:
パラメータ削除用メソッドを作成

gomo commented:
http://moriblog.kit-eng.com/?p=981

参照型のKeyValuePairのようなクラス。Tupleがあった。

https://msdn.microsoft.com/ja-jp/library/system.tuple(v=vs.100).aspx

どっちを使うのが良いのか？・・・
まあ、とりあえず後回しでいいかな。内部的なことだし。

sd-tomita committed:
特定のKeyに一致するValueの数を控えておく処理を追加

GetParams の List.ForEach 中に break する方法が
見つからなかったので、そこだけはとりあえず foreach ステートメントを
使用しました。

sd-tomita committed:
参照型のクラス Tuple の List でパラメータを管理するように変更

より安全と思しきため変更
構造体 KeyValuePair よりもメモリ使用量も少なく済む　はず

sd-tomita committed:
空のDictionaryが渡された場合に備えた処理を追記

実際はあまりそういう使い方はされないと思われるが、念のため追加

sd-tomita committed:
Tuple の Item1, Item2 が何かをコメントで追記

どっちがどっちだかわからんｗ

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -28,26 +24,28 @@ public Url(string urlStr)
-            string[] tmp = str.Split('=');
-            this.paramData[tmp[0]] = tmp[1];
-          });
+          QueryStringList = query.Split('&').ToList();
```
この時

``` c#
var QueryStringList = new List<string>();
```

が無駄になりそうな気がします。

``` c#
List<string> QueryStringList = null;

if(query.Contains('&'))
{

}
else
{
   QueryStringList = new List<string>();
   QueryStringList.Add(query);
}
```

みたいな感じで行けるかと

:innocent:

gomo replied:
あれ？先を読んで思ったけど、QueryStringListがそもそも必要かな？

``` c#
// uri.Query.Trim('?').query.Split('&').ForEach();ができたらそっち。
foreach(var keyValue in uri.Query.Trim('?').query.Split('&'))
{
   string[] tmp = str.Split('=');
   this.ParamList.Add(Tuple.Create(tmp[0], tmp[1]));
   this.AddRoopCount(tmp[0]);
}
```

で行けそうな気がした。Splitは対象文字がなかった時、そのものを配列として返しませんでしたっけ？`if(query.Contains('&'))`自体の判断もいらない気がしました。

また、微妙に別件ですが、`(string[]).ToList()`の計算量はO(n)で、結構重い処理です。必要なければ避けましょう。今回は逐次処理ができればいいだけなので必要ないですよね？

とりあえず、QueryStringListが廃止できたら廃止してください。で、どちらにしても`ToList`はしないで実装してください。

:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -7,17 +7,13 @@ namespace Sdx.Web
       var uri = new Uri(urlStr);
-      this.paramData = new Dictionary<string, string>();
+      this.ParamList = new List<Tuple<string,string>>();
+      this.RoopCount = new Dictionary<string, int>();
```
コンストラクタで初期化しなくても

``` c#
    private List<Tuple<string, string>> ParamList
    {
      get; set;
    } = new List<Tuple<string, string>>();
```

という若干気持ち悪い初期化が可能です。

``` c#
    private List<Tuple<string, string>> ParamList { get; set; } = new List<Tuple<string, string>>();
```

一行で書いたほうが気持ち悪くないかなｗ

どっちでやっても変わらないけど、タイミングが違うらしいですね・・それにより何が違うのかというと・・・
http://stackoverflow.com/questions/298183/c-sharp-member-variable-initialization-best-practice
↑ここをよく読むと分かるかもｗ私もまだよくわかってません。

まあ、出来るよということは知っておいてください。

ちなみにこれが出来るようになったのはC#6からですが、C#6の機能も.NET4で動くようですね。
https://ufcpp.wordpress.com/2014/11/30/c%E3%81%AE%E8%A8%80%E8%AA%9E%E3%83%90%E3%83%BC%E3%82%B8%E3%83%A7%E3%83%B3%E3%81%A8-net-framework%E3%83%90%E3%83%BC%E3%82%B8%E3%83%A7%E3%83%B3/

ちょっと気になったのは、ヌキなびがclassicで動いてるので、一応確認したほうが良さそうですね。確認しておきます。

これは直さなくてOKです。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -74,18 +70,24 @@ private string BuildPath()
-          .ToDictionary(p => p.Key, p => p.Value)
+        this.ParamList
+          .Concat(tpList)
+          .ToList()
```
Concatじゃなくて、こっちのほうが効率いいかも。
https://msdn.microsoft.com/ja-jp/library/z883w3dc(v=vs.100).aspx

:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -74,18 +70,24 @@ private string BuildPath()
+      }
+
+      var tpList = new List<Tuple<string, string>>();
+      tpList.Add(Tuple.Create(add.First().Key, add.First().Value));
```
これだと最初の要素しか足されないですね。あと、クエリーの順番を固定したいシチュエーションに対応するため`List<Tuple<string, string>>`にしたんですけど、ここがDictionaryだとここが保証されなくなってしまいますね。

これはどうするか考えます。とりあえず、複数追加できるようにしておいてください。

:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -105,62 +107,107 @@ public string Build(string[] exclude)
+          .ToList()
       );
       return path + query;
     }
```
同じ実装のメソッドが2つあるけど、これは`IEnumerable<string>`だとダメかな？`Contains`さえあればOKなんですよね？試してみてください。

:innocent:

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -105,62 +107,107 @@ public string Build(string[] exclude)
+    /// 同じ名前のキーがセットされたらカウントをアップし、
+    /// 削除されたらカウントを0に戻す
+    /// </summary>
+    private Dictionary<string, int> RoopCount
```
`RoopCount`という名前が若干分かりづらいと思いました。`ParamCount`か`EachCount`とか`CountByKey`とか・・・まあ、微妙ですかね。privateなんでそのままでOKです。

gomo commented:
https://github.com/SunriseDigital/cs-sdx/blob/fix/ApiAccessManagementLibrary/addTest/Sdx.Web/Url.cs#L185-L188
ここは例外が良いんですかね？

NULLを返す場合は、どうせその先で例外になるので、ここで投げちゃったほうが良さそうな気はするけど、空のリストならリストを返しちゃったほうがその後の扱いが良いと思いました。
:innocent:

gomo commented:
https://github.com/SunriseDigital/cs-sdx/blob/fix/ApiAccessManagementLibrary/addTest/Sdx.Web/Url.cs#L163-L166

こちらは例外のママが良いかな。

> NULLを返す場合は、どうせその先で例外になるので、ここで投げちゃったほうが良さそうな気はするけど

なので。ただ、なげるならチェックするメソッドも合わせて用意した方がいいですね。チェックメソッドもお願いします。

`ContainsKey`でいいかな・・・`ContainsParam`がいいか・・・Paramの方が統一感会っていいかな。

:innocent:

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -83,6 +165,173 @@ public void TestSingleParam()
+    }
+
+    [Fact]
+    public void TestNullCharValueParams()
```
細かいですけど、これはNullCharではなく、EmptyStringじゃないかな。

https://ja.wikipedia.org/wiki/%E7%A9%BA%E6%96%87%E5%AD%97%E5%88%97

Wiki見ると同じように書いてあるけど、ヌル文字って言うとC言語（だったかな？）で文字列の終端を表すヌル文字が浮かぶのでプログラマー的には微妙かと思いました。

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -83,6 +165,173 @@ public void TestSingleParam()
+      Assert.Equal("http://example.com/path/to/api?key=&foo=", url.Build());
+
+      //Null を Add
+      url.AddParam("key", null);
```
あ、この辺でNULLを使ってるから、その発想になったのかな？まあ、直さなくてもいいかな。公開インターフェースってわけじゃないですからね。いつでも変更できるし。

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -83,6 +165,173 @@ public void TestSingleParam()
+
+      //空文字を Add
+      url.AddParam("foo", "");
+      Assert.Equal("http://example.com/path/to/api?key=&foo=&key=&foo=", url.Build());
```
複数ある、空の値をもったパラメータの取得テストも追加お願いします。

:innocent:

sd-tomita commented:
諸々また別ブランチで修正します。
